sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java

```diff
@@ -171,6 +167,81 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
             addressSelector);
     }
 
+    private Mono<StoreResponse> getStoreResponseMonoWithSpeculation(RxDocumentServiceRequest request, Quadruple<Boolean, Boolean, Duration, Integer> forceRefreshAndTimeout) {
+        CosmosE2EOperationRetryPolicyConfig config = request.requestContext.getE2EOperationRetryPolicyConfig();
+        List<Mono<StoreResponse>> monoList = new ArrayList<>();
+        List<RxDocumentServiceRequest> requests = new ArrayList<>();
+
+        if (speculativeProcessor.getRegionsForPureExploration() != null
+            && !speculativeProcessor.getRegionsForPureExploration().isEmpty()) {
+            explore(request, forceRefreshAndTimeout);
+        }
+
+        if (speculativeProcessor.shouldIncludeOriginalRequestRegion()) {
+            monoList.add(getStoreResponseMono(request, forceRefreshAndTimeout));
+            requests.add(request);
+        }
+
+        for (URI endpoint : speculativeProcessor.getRegionsToSpeculate(config, this.transportClient
+            .getGlobalEndpointManager().getAvailableReadEndpoints())) {
+
+            if (request.requestContext.locationEndpointToRoute != endpoint) {
+                RxDocumentServiceRequest newRequest = request.clone();
+                newRequest.requestContext.routeToLocation(endpoint);
+                requests.add(newRequest);
+                monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout).delaySubscription(speculativeProcessor.getThreshold(config)));
+            }
+        }
+
+        return Mono.firstWithValue(monoList).map(storeResponse ->
+            {
+                for (RxDocumentServiceRequest r : requests) {
+                    CosmosDiagnostics diagnostics = r.requestContext.cosmosDiagnostics;
+                    // for the successful request region, we will use the actual latency
+                    if (request.getActivityId().toString().equals(storeResponse.getActivityId())) {
+                        speculativeProcessor.onResponseReceived(r.requestContext.locationEndpointToRoute,
+                            diagnostics.getDuration());
+                    }
+                }
+                return storeResponse;
+            });
+    }
+
+    private void explore(RxDocumentServiceRequest request, Quadruple<Boolean, Boolean, Duration, Integer> forceRefreshAndTimeout) {
+        for (URI endpoint : speculativeProcessor.getRegionsForPureExploration()) {
+            if (request.requestContext.locationEndpointToRoute != endpoint) {
+                RxDocumentServiceRequest newRequest = request.clone();
+                newRequest.requestContext.routeToLocation(endpoint);
+                getStoreResponseMono(newRequest, forceRefreshAndTimeout)
```

Comment:
Does it make sense to have a dedicated execution context / scheduler for exploration requests?